File: src/leetcode/76minWindow.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        vector<bool> valid(128, false);
        vector<int> freq(128, 0);
        for (int i = 0; i < t.length(); ++i) {
            valid[t[i]] = true;
            ++freq[t[i]];
        }

        int min_l = -1, min_length = -1, count = 0;
        for (int l = 0, r = 0; r < s.length(); ++r) {
            if (!valid[s[r]]) {
                continue;
            }
            if (--freq[s[r]] >= 0) {
                ++count;
            }
            while (count == t.length()) {
                if (min_length == -1 || r - l + 1 < min_length) {
                    min_l = l;
                    min_length = r - l + 1;
                }
                if (valid[s[l]] && ++freq[s[l]] > 0) {
                    --count;
                }
                ++l;
            }
        }
        return min_length == -1 ? "" : s.substr(min_l, min_length);
    }
};

int main(int argc, char *argv[])
{
    string s = "ADOBECODEBANC";
    string t = "ABC";
    cout << Solution().minWindow(s, t) << endl;
    return 0;
}

File: scripts/leetcode/76minWindow.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        freq = dict()
        for c in t:
            freq[c] = freq.get(c, 0) + 1
        min_l, min_length = None, None
        count = 0
        l = 0
        for r in range(len(s)):
            if s[r] not in freq:
                continue
            freq[s[r]] -= 1
            if freq[s[r]] >= 0:
                count += 1
            while count == len(t):
                if min_length == None or r - l + 1 < min_length:
                    min_l = l
                    min_length = r - l + 1
                # if s[l] in freq and freq[s[l]] + 1 > 0:
                #     count -= 1
                if s[l] in freq:
                    freq[s[l]] += 1
                    if freq[s[l]] > 0:
                        count -= 1
                l += 1
        return "" if min_length == None else s[min_l: min_l + min_length]
s = "ADOBECODEBANC"
t = "ABC"
print(f"results: {Solution().minWindow(s, t)}")